### PR TITLE
Week 04: For Home

### PR DESCRIPTION
**Pull request by pvarna**

### CODE REVIEW

File: Week04-ConvolutionalNeuralNetworks/task09.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from torch.utils.data import Dataset, DataLoader
+from torchvision.datasets import ImageFolder
+from torchvision import transforms
+
+
+class OmniglotDataset(Dataset):
+
+    def __init__(self, path_to_images, transformations):
+        super().__init__()
+        self.image_folder = ImageFolder(path_to_images, transformations)
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task09.py:15]: Implement the logic - do not use directly.

File: Week04-ConvolutionalNeuralNetworks/task09.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    def __len__(self):
+        return len(self.image_folder)
+
+    def __getitem__(self, index):
+        image_as_tensor, label = self.image_folder[index]
+
+        one_hot_encoding_tensor = torch.zeros(len(self.image_folder.classes))
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task09.py:23]: Use pd.get_dummies https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html

File: Week04-ConvolutionalNeuralNetworks/task09.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def main():
+    train_transforms = transforms.Compose(
+        [transforms.ToTensor(),
+         transforms.Resize((64, 64))])
+
+    dataset = OmniglotDataset("../DATA/omniglot_train", train_transforms)
+    size = dataset.__len__()
+    image_tensor, _, _ = last_element = dataset.__getitem__(size - 1)
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task09.py:36]: Never use methods beginning with two underscores. This means that there most likely an operator. In this case there is one: the square brackets, i.e. dataset[-1]

File: Week04-ConvolutionalNeuralNetworks/task09.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        return image_as_tensor, one_hot_encoding_tensor, label
+
+
+def main():
+    train_transforms = transforms.Compose(
+        [transforms.ToTensor(),
+         transforms.Resize((64, 64))])
+
+    dataset = OmniglotDataset("../DATA/omniglot_train", train_transforms)
+    size = dataset.__len__()
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task09.py:35]: len(dataset)

**Review by SimeonHristov99 — CHANGES_REQUESTED** [review]

File: Week04-ConvolutionalNeuralNetworks/task01.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        # https://pandas.pydata.org/docs/reference/api/pandas.Series.to_numpy.html
+        np_array = self.water_potability_data.iloc[index].to_numpy()
+        return np_array[:-1], np_array[-1]
+
+
+def main():
+    water_potability_dataset = WaterPotabilityDataset(
+        "../DATA/water_train.csv")
+
+    print(f"Number of instances: {water_potability_dataset.__len__()}")
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task01.py:24]: Use len(water_potability_dataset)

File: Week04-ConvolutionalNeuralNetworks/task01.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    print(f"Fifth item: {water_potability_dataset.__getitem__(4)}")
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task01.py:25]: water_potability_dataset[4]

File: Week04-ConvolutionalNeuralNetworks/task01.py (hunk 5 above)
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task01.py:24]: Avoid using methods beginning with two underscores.

File: Week04-ConvolutionalNeuralNetworks/task02.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                num_epochs=1000,
+                create_plot=True)
+
+    metric = torchmetrics.F1Score(task="binary", average="micro")
+    f1_score = evaluate_model(test_dataloader, net, metric)
+
+    print(f"F1 score on test set: {f1_score:.16f}")
+
+    # The model doesn't perform very well (F1 score is ~0.56)
+    # https://encord.com/blog/f1-score-in-machine-learning/ --> here it says it should be >0.8
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task02.py:163]: There is no definitive threshold. A lot of sources can say different numbers - what matters is what our clients / stakeholders want. Most of the time it would a model that is better than the one that they currently have / the one that the competition offers. So aim to be in the top 25%, but don't stress over it. As we say in the video from week 1, the only proven way to have better models is: more data, more layers, and more epochs.

File: Week04-ConvolutionalNeuralNetworks/task03.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        super(Net, self).__init__()
+        self.fc1 = nn.Linear(9, 16)
+        self.bn1 = nn.BatchNorm1d(16)
+        self.fc2 = nn.Linear(16, 8)
+        self.bn2 = nn.BatchNorm1d(8)
+        self.fc3 = nn.Linear(8, 1)
+
+        init.kaiming_uniform_(self.fc1.weight, nonlinearity="leaky_relu")
+        init.kaiming_uniform_(self.fc2.weight, nonlinearity="leaky_relu")
+        init.kaiming_uniform_(self.fc3.weight, nonlinearity="leaky_relu")
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task03.py:31]: Should be nonlinearity="sigmoid" right?

File: Week04-ConvolutionalNeuralNetworks/task04.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+LEARNING_RATE = 0.0005
+DROPOUT_PROBABILITY = 0.2
+
+class Net(nn.Module):
+
+    def __init__(self):
+        super(Net, self).__init__()
+        self.fc1 = nn.Linear(9, 128)
+        self.bn1 = nn.BatchNorm1d(128)
+        self.d1 = nn.Dropout(p=DROPOUT_PROBABILITY)
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task04.py:25]: Usually this is a parameter of the model (passed to the constructor).

File: Week04-ConvolutionalNeuralNetworks/task05.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+def get_all_cloud_images():
+    # https://realpython.com/get-all-files-in-directory-python/
+    clouds = pathlib.Path("../DATA/clouds")
+
+    return list(clouds.rglob("*.jpg"))
+
+def main():
+    clouds = get_all_cloud_images()
+
+    # https://stackoverflow.com/questions/46615554/how-to-display-multiple-images-in-one-figure
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task05.py:16]: We have examples of this in notes.md as well.

File: Week04-ConvolutionalNeuralNetworks/task06.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    image = to_pil(image_tensor.squeeze(0))
+    image.show()
+
+def train_epoch(dataloader_train, optimizer, net, criterion, i):
+    running_loss = 0.0
+
+    for inputs, targets in tqdm(dataloader_train, desc=f"Epoch {i+1}"):
+        optimizer.zero_grad()
+
+        # inputs = inputs.float()
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task06.py:65]: If not needed, we can remove these two lines.

File: Week04-ConvolutionalNeuralNetworks/task07.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+import time
+import torchmetrics
+from pprint import pprint
+
+NUMBER_CLASSES = 7
+LEARNING_RATE = 0.001
+BATCH_SIZE = 16
+EPOCHS = 20
+
+TRAIN_TRANSFORMS = transforms.Compose([
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task07.py:19]: These could be in `main`.

File: Week04-ConvolutionalNeuralNetworks/task11.py [hunk 13]
```diff
@@ -2,10 +2,297 @@
...
+    for images, alphabet_labels, character_labels in tqdm(dataloader_train):
+        optimizer.zero_grad()
+
+        characters_outputs, alphabets_outputs = net(images)
+        characters_loss = characters_criterion(characters_outputs,
+                                               character_labels)
+        alphabets_loss = alphabets_criterion(alphabets_outputs,
+                                             alphabet_labels)
+
+        loss = characters_loss + alphabets_loss
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task11.py:109]: We could also weigh them

File: Week04-ConvolutionalNeuralNetworks/task11.py [hunk 14]
```diff
@@ -2,10 +2,297 @@
...
+                                       batch_size=BATCH_SIZE,
+                                       shuffle=True)
+    dataloader_validation = data.DataLoader(dataset_validation,
+                                            batch_size=BATCH_SIZE,
+                                            shuffle=False)
+
+    net = OmniglotTwoOutputNet()
+    optimizer = optim.AdamW(net.parameters(), lr=LEARNING_RATE)
+    characters_f1_metric = torchmetrics.F1Score(task="multiclass",
+                                                average="macro",
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task11.py:282]: micro

File: Week04-ConvolutionalNeuralNetworks/task11.py [hunk 15]
```diff
@@ -2,10 +2,297 @@
...
+                                                num_classes=CHARACTERS_COUNT)
+    alphabets_f1_metric = torchmetrics.F1Score(task="multiclass",
+                                               average="macro",
```
Comment (SimeonHristov99) [inline Week04-ConvolutionalNeuralNetworks/task11.py:285]: micro - better for unbalanced datasets

**Review by SimeonHristov99 — APPROVED** [review]
